internal/handlers/common/error.go

```diff
@@ -108,21 +139,88 @@ func (e *Error) Document() *types.Document {
 	return d
 }
 
-// ProtocolError converts any error to wire protocol error.
-//
-// Nil panics, *Error (possibly wrapped) is returned unwrapped with true,
-// any other value is wrapped with InternalError and returned with false.
-func ProtocolError(err error) (*Error, bool) {
-	if err == nil {
-		panic("err is nil")
+// WriteErrors represents slice of protocol write errors.
+type WriteErrors []WriteError
+
+// NewWriteErrorMsg creates new protocol write error with given ErrorCode and message.
+func NewWriteErrorMsg(code ErrorCode, msg string) error {
+	return &WriteErrors{{
+		code: code,
+		err:  errors.New(msg),
+	}}
+}
+
+// Error implements error interface.
+func (we *WriteErrors) Error() string {
+	var err string
+	for _, e := range *we {
+		err += e.Error() + ","
 	}
 
-	var e *Error
-	if errors.As(err, &e) {
-		return e, true
+	return err
+}
+
+// Code implements ProtoErr interface.
+func (we *WriteErrors) Code() ErrorCode {
+	for _, e := range *we {
+		return e.code
 	}
+	return errUnset
+}
 
-	return NewError(errInternalError, err).(*Error), false
+// Unwrap implements standard error unwrapping interface.
+func (we *WriteErrors) Unwrap() error {
+	for _, e := range *we {
+		return &e
+	}
+	return nil
+}
+
+// Document implements ProtoErr interface..
```

Comment:
Could you please add here some more from your debugging adventures?

Reply:
Added.